src/internal_modules/roc_audio/latency_tuner.cpp

```diff
@@ -422,18 +562,178 @@ void LatencyTuner::report_() {
+                (long)link_metrics_.lost_packets,
+                (double)link_metrics_.jitter / core::Millisecond,
+                (double)link_metrics_.max_jitter / core::Millisecond,
+                (double)link_metrics_.min_jitter / core::Millisecond,
+                (long)link_metrics_.total_packets);
+        roc_log(LogDebug, "Latency monitor: fec block duration=%.1fms",
+                (double)latency_metrics_.fec_block_duration / core::Millisecond);
+    }
+
+    if (sample_spec_.ns_2_stream_timestamp_delta(latency_metrics_.fec_block_duration)
+        >= max_latency_) {
+        roc_log(LogInfo,
+                "Latency tuner: FEC block %.1fms is longer than the max "
+                "limit for latency %d(%.1fms)",
+                (double)latency_metrics_.fec_block_duration / core::Millisecond,
+                max_latency_, (double)max_latency_ / core::Millisecond);
+    }
+}
+
+// Decides if the latency should be adjusted and orders fe_ to do so if needed.
+//
+// 1. Decides to decrease latency if current value is greater than upper threshold,
+//    The target latency is supposed to change smoothely, so we just cut the current
+//    latency value by some percentage.
+//
+// 2. Decides to increase latency if it is lesser than lower threshold (which
+//    could be close or equal to target latency itself).
+//    This could/should be done effectively as it could possibly mean that the user
+//    is already perceives some losses.
+//
+//    NB: After the increasement the new latency target value must not be greater than
+//        upper threshold in any circumstances.
+//
+//
+void LatencyTuner::update_target_latency_(const core::nanoseconds_t max_jitter_ns,
+                                          const core::nanoseconds_t mean_jitter_ns,
+                                          const core::nanoseconds_t fec_block_ns) {
+    const core::nanoseconds_t estimate = std::max(
+        std::max((core::nanoseconds_t)(max_jitter_ns * 1.15), mean_jitter_ns * 3),
+        fec_block_ns);
+    const core::nanoseconds_t now = core::timestamp(core::ClockMonotonic);
+    core::nanoseconds_t cur_tl_ns =
+        sample_spec_.stream_timestamp_delta_2_ns(target_latency_);
+
+    if (target_latency_state_ == TL_NONE) {
+        // If there is no active timeout, check if evaluated target latency is
+        // significantly smaller, than the latency in action so that we could decrease it.
+        if (estimate < cur_tl_ns && estimate * lat_update_upper_thrsh_ < cur_tl_ns
+            && fe_->stable()) {
+            const core::nanoseconds_t new_tl_ns =
+                (core::nanoseconds_t)(cur_tl_ns * lat_update_dec_step_);
+            const packet::stream_timestamp_diff_t new_tl_ts =
+                sample_spec_.ns_2_stream_timestamp_delta(new_tl_ns);
+            if (new_tl_ts < min_latency_) {
+                if (now > last_lat_limit_log_
+                    && (now - last_lat_limit_log_) > 5 * core::Second) {
```

Comment:
Please move all hard-coded timeouts like 5 seconds, 15 seconds, to LatencyConfig struct. (Also please add a comment to explain their meaning)